src/room007/models/word_tag_predictor.py

```diff
@@ -15,117 +17,130 @@
-            lambda row: [1] * len(row['title_non_stop_words']) + [0] * len(row['content_non_stop_words']),
-            axis=1)))
-
-
-    def _add_number_of_non_stop_words(self, data):
-        split_row = lambda row: [x for x in row.split() if x not in stop_words]
-        data['title_non_stop_words'] = data['title'].apply(split_row)
-        data['content_non_stop_words'] = data['content'].apply(split_row)
-
-
-    def _is_in_question(self, data):
-        def is_in_question(row):
-            features = []
-            question = 0
-            for word in row.split()[::-1]:
-                if word in '.:;!?':
-                    question = int(word == '?')
-                if word not in stop_words:
-                    features.append(question)
-            return features[::-1]
-        return list(itertools.chain(*data['titlecontent'].apply(
-            is_in_question
-            )))
-
-    def _times_word_in(self, data, column):
-        return list(itertools.chain(*data.apply(
-            lambda row: [row[column].split().count(word)
-                         for word in row['titlecontent'].split()
-                         if word not in stop_words], axis=1)
-            ))
+    def _get_features_per_row(self, data):
+        features = self._get_data_independent_features(data)
+        features.append(self._get_tf_idf_features_per_word(data))
+        #if self.changes:
+        #    s_feats = [self._title_or_content(data)]
+        #    s_feats = self._spacy_features(data)
+        #    features.extend(s_feats)
+        return features
 
+    def _get_data_independent_features(self, data):
+        _add_number_of_non_stop_words(data)
+        return [
+                _times_word_in(data, 'title'),
+                _times_word_in(data, 'content'),
+                _is_in_question(data),
+                _title_or_content(data),
+        ]
+
+    def _spacy_features(self, data):
+        return [data.apply(
+            lambda row: [int(tags.pos_ == "NOUN") for word, tags in zip(row['titlecontent'].split(), nlp(row['titlecontent'])) if word not in stop_words],
+            axis=1)]
 
     def _train_tf_idf_vectorizer(self, data):
         self.tf_idf_vectorizer = TfidfVectorizer(stop_words=stop_words)
         self.tf_idf_vectorizer.fit(data['titlecontent'])
         #if self.functional_test:
         #    self._write_example_it_idf_features(data)
 
-
     def _get_tf_idf_features_per_word(self, train_data):
         tf_idf_data = self.tf_idf_vectorizer.transform(train_data['titlecontent'])
         train_data['index'] = range(tf_idf_data.shape[0])
         voc = self.tf_idf_vectorizer.vocabulary_
-        features = list(itertools.chain(*train_data.apply(
+        features = train_data.apply(
             lambda row: [tf_idf_data[row['index'], voc.get(word)]
-                         if voc.get(word) else (0 if not self.changes else 1)
+                         if voc.get(word) else 1
```

Comment:
Well the whole advanced_train_and_predict is to make use of this self.changes, otherwise you can just use the train_and_predict. 
Because the question we want to answer is this:
If I change something, (for example adding code), what is the effect on my result. (and this needs to be possible on a random subset of the data without having to things manually)
So in this moment, the change variable has no use, but that is just because the current changes don't involve a change to the 'recognition' algorithm ( if I hadn't commented out those spacy things then the self.changes has it uses).